pkg/controllers/machinesetmigration/machineset_migration_controller.go

```diff
@@ -0,0 +1,391 @@
+			return fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		if annotations.HasPaused(capiMachineSet) {
+			capiMachineSetCopy := capiMachineSet.DeepCopy()
+			delete(capiMachineSet.Annotations, capiv1beta1.PausedAnnotation)
+
+			if err := r.Patch(ctx, capiMachineSet, client.MergeFrom(capiMachineSetCopy)); err != nil {
+				return fmt.Errorf("failed to patch CAPI machine set: %w", err)
+			}
+		}
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// For requesting unpausing of a MAPI resource, it is sufficient to switch the spec.AuthoritativeAPI field on the MAPI resource.
+		// which is already done before this code runs in this controller. Nothing to do here.
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	return nil
+}
+
+func (r *MachineSetMigrationReconciler) ensurePausedRequestedOnOldAuthoritativeAPI(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) error {
+	// Request that the old authoritative resource reconciliation is paused.
+	switch mapiMachineSet.Spec.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		// For requesting pausing of a MAPI resource, it is sufficient to switch the spec.AuthoritativeAPI field on the MAPI resource.
+		// which is already done before this code runs in this controller.
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		// For requesting pausing of a CAPI resource, set the paused annotation on it.
+		// The spec.AuthoritativeAPI is set to MachineAPI, meaning that the old authoritativeAPI was ClusterAPI.
+		// So Check if the ClusterAPI resource has the paused annotation, otherwise set it.
+		capiMachineSet := &capiv1beta1.MachineSet{}
+		if err := r.Get(ctx, client.ObjectKey{Namespace: consts.DefaultManagedNamespace, Name: mapiMachineSet.Name}, capiMachineSet); err != nil {
+			return fmt.Errorf("failed to get CAPI machine set: %w", err)
+		}
+
+		if !annotations.HasPaused(capiMachineSet) {
+			capiMachineSetCopy := capiMachineSet.DeepCopy()
+			annotations.AddAnnotations(capiMachineSet, map[string]string{capiv1beta1.PausedAnnotation: ""})
+			if err := r.Patch(ctx, capiMachineSet, client.MergeFrom(capiMachineSetCopy)); err != nil {
+				return fmt.Errorf("failed to patch CAPI machine set: %w", err)
+			}
+		}
+	default:
+		// Any other value is disallowed by the openAPI schema validation.
+	}
+
+	return nil
+}
+
+// getConditionStatus returns the status for the condition.
+func getConditionStatus(mapiMachineSet *machinev1beta1.MachineSet, condition machinev1beta1.ConditionType) corev1.ConditionStatus {
+	for _, c := range mapiMachineSet.Status.Conditions {
+		if c.Type == condition {
+			return c.Status
+		}
+	}
+
+	return corev1.ConditionUnknown
+}
+
+func (r *MachineSetMigrationReconciler) isSynchronizedGenerationMatchingAuthoritativeAPIGeneration(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet) (bool, error) {
+	// Request that the old authoritative resource reconciliation is paused.
```

Comment:
Might want to explain why the checks appear reversed initially, took me a moment to realise

Reply:
Yes, done.